### PR TITLE
Introduce Sphinx-backed developer documentation

### PR DESCRIPTION
**Pull request by lunaris**
This commit introduces an initial Sphinx configuration for writing repository-wide developer documentation in Sphinx's MyST Markdown. It includes Sphinx and MyST configuration as well as workflows for local development. Documentation on writing documentation -- conventions as well as how to best utilise Sphinx and other tools is included as a starting point, with the intention being to fill out the remainder of the repository in subsequent commits (this will include migrating and deprecating `developer-docs`). Deployment uses the existing "Read the docs" webhook and will overwrite the current live version, which at this point is 6 months' stale anyway.

### CODE REVIEW

**Comment by lunaris** [comment]
An example build of this PR has been manually deployed to https://pulumi-developer-docs.pages.dev/ for people to take a look/get a feel for how this will appear.

File: docs/documentation.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+### Building
+
+The `docs` directory contains a `Makefile` (<gh-file:pulumi#docs/Makefile>) for
+working with the documentation. Generally you'll just want:
+
+```sh
+make -C docs watch
+```
+
+which will watch for changes and rebuild the documentation as needed, serving
```
Comment (Frassle) [inline docs/documentation.md:20]: ❤️

File: docs/writing.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
+# Writing documentation
+
+## Conventions
+
+* Write documentation in Markdown (`.md` files) where possible. While Sphinx
```
Comment (Frassle) [inline docs/writing.md:5]: To be simple can we just remove all the rst comments in this doc? Maybe just note rst is supported by sphinx but we're not using it, then just focus on markdown.

**Comment by Frassle** [comment]
>  eventually deploying to ReadTheDocs to replace the old documentation, using GitHub Pages (with or without a custom domain), or using something else (e.g. Cloudflare Pages).

What's simplest?

**Comment by lunaris** [comment]
> > eventually deploying to ReadTheDocs to replace the old documentation, using GitHub Pages (with or without a custom domain), or using something else (e.g. Cloudflare Pages).
> 
> What's simplest?

I feel like GitHub Pages would be, but we have an org-wide custom domain setup (pulumi.com) meaning that any GH Pages project we create must either live under pulumi.com/<repo> or have an overriding custom domain setup. Since I don't have a domain to spare/am not versed in how we configure those, I have for now set up a Cloudflare Pages project and am manually pushing to that. CF Pages is just as simple, just takes us outside the GH ecosystem a bit (though we already have accounts and the like there for testing providers, etc.)

**Comment by lunaris** [comment]
Update: I have wired this into the existing (stale) Read the docs workflow for now -- see https://pulumi-developer-docs--16953.org.readthedocs.build/16953/ for this PR

**Review by julienp — APPROVED** [review]

**Pull request merged**